### PR TITLE
util: Use locale independent ToString(…) instead of locale dependent strprintf(…) for low-level string formatting

### LINKED ISSUE #18281: Low level functions ScriptToAsmStr (core_io), i64tostr (strencodings) and itostr (strencodings) are locale dependent
Low level functions `ScriptToAsmStr` (`core_io`), `i64tostr` (`strencodings`) and `itostr` (`strencodings`) are locale dependent:

```c++
const std::vector<uint8_t> b{0x6a, 0x4, 0xff, 0xff, 0xff, 0xff};
const CScript script{b.begin(), b.end()};
for (const std::string& locale_string : {"C", "de_DE"}) {
    std::locale::global(std::locale(locale_string));
    std::cout << "[" << locale_string << "] ScriptToAsmStr(script, false) == " 
        << ScriptToAsmStr(script, false) << "\n";
    std::cout << "[" << locale_string << "] i64tostr(12345678) == " 
        << i64tostr(12345678) << "\n";
    std::cout << "[" << locale_string << "] itostr(12345678) == " 
        << itostr(12345678) << "\n";
}
```

```
[C] ScriptToAsmStr(script, false) == OP_RETURN -2147483647
[C] i64tostr(12345678) == 12345678
[C] itostr(12345678) == 12345678
[de_DE] ScriptToAsmStr(script, false) == OP_RETURN -2.147.483.647
[de_DE] i64tostr(12345678) == 12.345.678
[de_DE] itostr(12345678) == 12.345.678
```

### PR DESCRIPTION
**Pull request by practicalswift**
Use locale independent `ToString(…)` instead of locale dependent `strprintf(…)` for low-level string formatting.

Context: See sipas comment in https://github.com/bitcoin/bitcoin/pull/18147#issuecomment-604727148.

Example taken from #18281:

> Low level functions `ScriptToAsmStr` (`core_io`), `i64tostr` (`strencodings`) and `itostr` (`strencodings`) are locale dependent:
> 
> ```c++
> const std::vector<uint8_t> b{0x6a, 0x4, 0xff, 0xff, 0xff, 0xff};
> const CScript script{b.begin(), b.end()};
> for (const std::string& locale_string : {"C", "de_DE"}) {
>     std::locale::global(std::locale(locale_string));
>     std::cout << "[" << locale_string << "] ScriptToAsmStr(script, false) == " 
>         << ScriptToAsmStr(script, false) << "\n";
> }
> ```
> 
> ```
> [C] ScriptToAsmStr(script, false) == OP_RETURN -2147483647
> [de_DE] ScriptToAsmStr(script, false) == OP_RETURN -2.147.483.647
> ```

Fixes #18281.

### CODE REVIEW

**Comment by practicalswift** [comment]
Oh, saw #18449 after submitting this. Closing!

**Comment by practicalswift** [comment]
Will rebase on top of #18449 and keep remaining changes.

**Comment by laanwj** [comment]
`strprintf` should not be locale dependent in the first place. It's never acceptable for it to be locale dependent. I think this is the wrong fix.

**Comment by practicalswift** [comment]
@laanwj I'm not sure I follow TBH.

I'm not sure if you are claiming that 1.) `tinyformat.h` (`strprintf`) isn't locale dependent, 2.) that you _wish_ that `tinyformat.h` wasn't locale dependent, or 3.) if you are suggesting that we should assert that the global C++ locale is set in a way that `strprintf` isn't locale dependent in practice? :)

FWIW:

```
$ cat tinyformat-locale.cpp
#include "tinyformat.h"

#include <iostream>
#include <locale>

int main(void)
{
    std::cout << strprintf("money printer go %d brrr", 1000000) << std::endl;
    std::locale::global(std::locale("de_DE"));
    std::cout << strprintf("gelddrucker gehen %d brrr", 1000000) << std::endl;
}
$ clang++ -o tinyformat-locale tinyformat-locale.cpp
$ ./tinyformat-locale
money printer go 1000000 brrr
gelddrucker gehen 1.000.000 brrr
```

**Comment by practicalswift** [comment]
@laanwj Also, is the claim about "wrong fix" referring to the tinyformat fix or all fixes in this PR? Please clarify - the review was a bit vague :)

**Comment by practicalswift** [comment]
Please also see #18432 which makes `tinyformat.h` (`strprintf`) locale independent. Perhaps that is the preferred fix?

I don't have any preference myself: I just want to kill this bug class once and for all :)

**Comment by MarcoFalke** [comment]
This needs rebase either way

**Comment by practicalswift** [comment]
@MarcoFalke Thanks! Rebased :)

File: src/net.cpp [hunk 1]
```diff
@@ -7,19 +7,19 @@
 #include <config/bitcoin-config.h>
 #endif
 
-#include <net.h>
```
Comment (MarcoFalke) [inline src/net.cpp:10]: the header of the own module is always included first on its own section

**Comment by practicalswift** [comment]
@laanwj Friendly ping: could you please clarify your review feedback? :)

**Review by luke-jr — APPROVED** [review]
utACK

**Comment by MarcoFalke** [comment]
ACK 4cb9397088bf01bb5893bbcb705668894da118a6

**Comment by jonatack** [comment]
Code review ACK 4cb9397088bf01bb5893bbcb705668894da118a6

**Review by jonatack — COMMENTED** [review]
Code review ACK 4cb9397

**Comment by practicalswift** [comment]
Re-opened thanks to renewed interest in the form of @jonatack's review and ACK :)

Unfortunately I had to rebase to resolve a conflict, so the three ACKs collected are now stale.

When rebasing I also added this case which makes this change complete across the code base:

```diff
diff --git a/src/wallet/rpcwallet.cpp b/src/wallet/rpcwallet.cpp
index dda00f1fe..c6a6a7e8d 100644
--- a/src/wallet/rpcwallet.cpp
+++ b/src/wallet/rpcwallet.cpp
@@ -4222,7 +4222,7 @@ static UniValue upgradewallet(const JSONRPCRequest& request)
         "\nUpgrade the wallet. Upgrades to the latest version if no version number is specified\n"
         "New keys may be generated and a new wallet backup will need to be made.",
         {
-            {"version", RPCArg::Type::NUM, /* default */ strprintf("%d", FEATURE_LATEST), "The version number to upgrade to. Default is the latest wallet version"}
+            {"version", RPCArg::Type::NUM, /* default */ ToString(FEATURE_LATEST), "The version number to upgrade to. Default is the latest wallet version"}
         },
         RPCResults{},
         RPCExamples{
```

@luke-jr @MarcoFalke @jonatack - would you mind re-reviewing the rebased version and re-ACK if everything still looks good? :)

**Comment by practicalswift** [comment]
@laanwj Friendly ping: can you clarify your review feedback as requested above?

Generally I don't think it is fair to leave NACK style comments if one is unwilling to follow-up when asked for clarification regarding the rationale.

**Pull request closed without merging**